File: verdict.js
/*
 * verdict.js — the VERDICT contract (CORE): `tepna.verdict/1`
 * Copyright 2026 Michal Planicka
 * SPDX-License-Identifier: Apache-2.0
 *
 * Licensed under the Apache License, Version 2.0 (the "License"); you may not
 * use this file except in compliance with the License. See the LICENSE and
 * NOTICE files at the project root, or http://www.apache.org/licenses/LICENSE-2.0
 *
 * ────────────────────────────────────────────────────────────────────────
 * Executes `VERDICT-CONTRACT-2026-09-21-BRIEF.md` §2 (owner standing requirement, CLAUDE.md §🧾):
 * every gate, oracle, audit, harness or study that DECIDES something emits ONE JSON object of a fixed
 * shape beside its prose. Prose is explanation; the object is the API. This module is the ONLY
 * definition of that shape — the enum, the field rules, the invariants — the way `measurement-block.js`
 * is the only definition of a measurement instance. A tool that hand-writes `"status": "PASSED"` reds
 * HERE, not in a reader's regex.
 *
 * THE FAILURE CLASS: three of the seven states a gate can be in — NOT_RUN, NOT_APPLICABLE, UNDERPOWERED —
 * read as green to a regex looking for "FAIL". A reader that must parse a paragraph to learn whether
 * evidence is trustworthy is §4b's "reported success about something it never examined" one layer up.
 *
 * `validate(obj) → { ok, errors[], checked[] }` — `checked` publishes which legs RAN (the denominator),
 * so a leg that could not run says so instead of passing silently. No dependencies; classic script;
 * loads in Node (vm co-load, `require`) and the browser suite alike.
 * ──────────────────────────────────────────────────────────────────────── */
(function (root) {
  'use strict';

  var SCHEMA = 'tepna.verdict/1';
  /* THE closed enum — exactly seven, stated once. Asserted as an equality by the gate group. */
  var STATUSES = ['PASS', 'FAIL', 'SHORTFALL', 'UNDERPOWERED', 'NOT_RUN', 'NOT_APPLICABLE', 'UNKNOWN'];
  var DIRECTIONS = ['lte', 'gte', 'eq', 'within'];
  /* P5 rides in the object: only a `publishable` verdict may be quoted outside the repo. Two values, closed. */
  var SCOPES = ['internal', 'publishable'];
  var SHA_RE = /^[0-9a-f]{7,40}$/;
  var ISO_UTC_RE = /^\d{4}-\d{2}-\d{2}T\d{2}:\d{2}:\d{2}(?:\.\d+)?Z$/;

  function _isObj(v) {
    return !!v && typeof v === 'object' && !Array.isArray(v);
  }
  function _isInt(v) {
    return typeof v === 'number' && Number.isFinite(v) && Math.floor(v) === v && v >= 0;
  }
  function _nonEmpty(s) {
    return typeof s === 'string' && s.trim().length > 0;
  }

  /**
   * validate(obj) → { ok, errors[], checked[] }
   *
   * Each leg pushes its name onto `checked` when it RAN; a leg that cannot run (the object is not
   * even an object, a sub-field is absent) records the absence as an error rather than skipping
   * silently. The invariants, in order: shape · status enum · population equality · criterion
   * pre-stated · result/reason rules per status · evidence · provenance.
   */
  function validate(obj) {
    var errors = [];
    var checked = [];
    var push = function (m) {
      errors.push(m);
    };
    if (!_isObj(obj)) {
      return {
        ok: false,
        errors: [typeof obj === 'string' ? 'a verdict is an OBJECT, not prose — a status word in a string is exactly what this contract forbids' : 'verdict is not an object'],
        checked: []
      };
    }

    // ── schema + gate ────────────────────────────────────────────────────────────────────────
    checked.push('schema');
    if (obj.schema !== SCHEMA) push('schema must be exactly "' + SCHEMA + '", got ' + JSON.stringify(obj.schema));
    if (!_nonEmpty(obj.gate)) push('gate must name the gate (non-empty string)');

    // ── status: the closed enum, case-exact ──────────────────────────────────────────────────
    checked.push('status enum');
    var st = obj.status;
    if (STATUSES.indexOf(st) < 0) {
      if (typeof st === 'string' && STATUSES.indexOf(st.toUpperCase()) >= 0) push('status "' + st + '" is a capitalisation variant — the enum is case-exact: ' + STATUSES.join('|'));
      else push('status must be one of ' + STATUSES.join('|') + ', got ' + JSON.stringify(st) + ' (no eighth value, no synonyms)');
    }

    // ── scope: internal | publishable — required; absence does not default (a verdict that has not
    //    said whether it may be quoted has not decided) ─────────────────────────────────────────
    checked.push('scope');
    if (SCOPES.indexOf(obj.scope) < 0) push('scope must be one of ' + SCOPES.join('|') + ' (P5: only publishable may be quoted outside the repo), got ' + JSON.stringify(obj.scope));

    // ── population: three integers, checked + excluded = eligible, an EQUALITY ──────────────
    var pop = obj.population;
    if (!_isObj(pop)) push('population missing — a verdict states what it examined: {checked, eligible, excluded}');
    else {
      checked.push('population equality');
      ['checked', 'eligible', 'excluded'].forEach(function (k) {
        if (!_isInt(pop[k])) push('population.' + k + ' must be a non-negative integer, got ' + JSON.stringify(pop[k]));
      });
      if (_isInt(pop.checked) && _isInt(pop.eligible) && _isInt(pop.excluded) && pop.checked + pop.excluded !== pop.eligible)
        push('population must satisfy checked + excluded = eligible (an equality, not a floor): ' + pop.checked + ' + ' + pop.excluded + ' ≠ ' + pop.eligible);
      if (st === 'PASS' && pop.checked === 0) push('PASS over population.checked = 0 is the examined-nothing shape — refused at the type level');
    }

    // ── criterion: named, pre-stated, with unit + direction ──────────────────────────────────
    var cr = obj.criterion;
    if (!_isObj(cr)) push('criterion missing — the rule the verdict applied, written BEFORE the measurement');
    else {
      checked.push('criterion');
      if (!_nonEmpty(cr.name)) push('criterion.name must be a non-empty string');
      if (DIRECTIONS.indexOf(cr.direction) < 0) push('criterion.direction must be one of ' + DIRECTIONS.join('|') + ', got ' + JSON.stringify(cr.direction));
      if (cr.direction === 'within') {
        if (!Array.isArray(cr.threshold) || cr.threshold.length !== 2 || !cr.threshold.every(Number.isFinite)) push('criterion.threshold for direction "within" must be a finite [lo, hi] pair');
      } else if (typeof cr.threshold !== 'number' || !Number.isFinite(cr.threshold)) push('criterion.threshold must be a finite number, got ' + JSON.stringify(cr.threshold));
      if (typeof cr.unit !== 'string') push('criterion.unit must be a string (use "" for a dimensionless count, never omit it)');
    }

    // ── result / reason, per status ──────────────────────────────────────────────────────────
    checked.push('result/reason per status');
    var hasResult = obj.result !== null && obj.result !== undefined;
    if (hasResult && !_isObj(obj.result)) push('result must be an object of measured quantities, or null');
    var reasonOk = _nonEmpty(obj.reason);
    if (st === 'PASS') {
      if (!hasResult) push('PASS requires result — the measured quantities the criterion was applied to');
      if (obj.reason !== null && obj.reason !== undefined) push('PASS must carry reason: null — a PASS that needs explaining is not a PASS');
    } else if (st === 'FAIL' || st === 'SHORTFALL') {
      if (!hasResult) push(st + ' requires result — what was measured');
      if (!reasonOk) push(st + ' requires reason — ' + (st === 'FAIL' ? 'what missed, by how much' : 'which sub-population or tail missed'));
    } else if (st === 'UNDERPOWERED') {
      if (!reasonOk) push('UNDERPOWERED requires reason naming the pre-stated minimum and the count');
      else if (!/\d/.test(obj.reason)) push('UNDERPOWERED reason must name the minimum and the count as NUMBERS, got "' + obj.reason + '"');
    } else if (st === 'NOT_RUN' || st === 'NOT_APPLICABLE') {
      if (hasResult) push(st + ' must carry result: null — ' + (st === 'NOT_RUN' ? 'nothing was examined, so nothing was measured' : 'the criterion does not bind, so no result exists'));
      if (!reasonOk) push(st + ' requires reason — ' + (st === 'NOT_RUN' ? 'why it did not run' : 'the property that makes the criterion inapplicable'));
    } else if (st === 'UNKNOWN') {
      if (!reasonOk) push('UNKNOWN requires reason — what could not be decided, and why');
    }

    // ── evidence: what a reader opens to re-derive the verdict ───────────────────────────────
    checked.push('evidence');
    if (!Array.isArray(obj.evidence)) push('evidence must be an array of paths/globs/records a reader would open');
    else {
      if (!obj.evidence.every(_nonEmpty)) push('evidence entries must be non-empty strings');
      if ((st === 'PASS' || st === 'FAIL' || st === 'SHORTFALL') && obj.evidence.length === 0) push(st + ' with empty evidence is a claim with nothing to open');
    }

    // ── provenance of the RUN (a real UTC instant — not a floating recording time) ───────────
    checked.push('provenance');
    var pb = obj.producedBy;
    if (!_isObj(pb)) push('producedBy missing — {tool, commit}');
    else {
      if (!_nonEmpty(pb.tool)) push('producedBy.tool must name the tool');
      if (pb.commit !== null && !SHA_RE.test(String(pb.commit || ''))) push('producedBy.commit must be a 7–40 hex sha, or null with the reason in `commitReason` when the tree is not a checkout');
      if (pb.commit === null && !_nonEmpty(pb.commitReason)) push('producedBy.commit is null without producedBy.commitReason (∅: say why)');
    }
    if (!ISO_UTC_RE.test(String(obj.at || ''))) push('at must be a real UTC instant, ISO-8601 with Z — this is provenance of the RUN, not a floating recording time');

    return { ok: errors.length === 0, errors: errors, checked: checked };
  }

  /**
   * make(fields) → a verdict object with the schema, `at`, and null defaults filled — a convenience for
   * producers so the shape is written once, here. Does NOT validate; call validate() on the result.
   */
  function make(f) {
    f = f || {};
    return {
      schema: SCHEMA,
      gate: f.gate,
      status: f.status,
      scope: f.scope === undefined ? 'internal' : f.scope, // the restrictive default; a producer WRITES publishable to lift it
      population: f.population,
      criterion: f.criterion,
      result: f.result === undefined ? null : f.result,
      evidence: f.evidence || [],
      reason: f.reason === undefined ? null : f.reason,
      producedBy: f.producedBy,
      at: f.at || new Date().toISOString().replace(/\.\d{3}Z$/, 'Z')
    };
  }

  var API = { SCHEMA: SCHEMA, STATUSES: STATUSES.slice(), SCOPES: SCOPES.slice(), DIRECTIONS: DIRECTIONS.slice(), validate: validate, make: make };
  root.Verdict = API;
  if (typeof module !== 'undefined' && module.exports) module.exports = API;
})(typeof globalThis !== 'undefined' ? globalThis : this);

File: capture-host/verdict.py
# tepna-capture — verdict.py
# Copyright 2026 Michal Planicka · SPDX-License-Identifier: Apache-2.0
"""`tepna.verdict/1` — a gate answers in ONE fixed shape; prose is explanation, not the API.

VERDICT-CONTRACT-2026-09-21 §1, owner 2026-09-21: *"a downstream machine cannot reliably distinguish
PASS / FAIL / NOT RUN / NOT APPLICABLE / UNDERPOWERED / SHORTFALL / UNKNOWN without parsing prose. That
is dangerous."* The box-side nightly gates are the first thing a clinician or the sealed-night reader
will ever see, so they emit the object beside the file they already write.

This module is the Python half of the contract: `make()` builds an object, `validate()` refuses a
malformed one (the same rules `verdict.js` applies in the JS lane), `write()` puts it beside a report
atomically. The rules that matter most are the ones that refuse the examined-nothing shape:

  · `status` is one of EXACTLY seven values — no synonyms, no case variants.
  · `population` is three integers with `checked + excluded == eligible` (an equality, not a floor —
    memory `gate-must-publish-its-denominator`).
  · a PASS with `checked == 0`, an empty `evidence`, or a non-null `reason` is invalid: that is a pass
    about nothing, refused at the type level. Every other status REQUIRES a reason.
  · `producedBy.commit` is a 7–40 hex sha, or null WITH a `commitReason` (absence says why).
  · `criterion` names the rule, threshold, unit and direction, and it was written before the
    measurement (memory `pre-state-the-threshold`) — the builders in nightqc.py carry theirs as
    module constants for exactly that reason.

`at` is a REAL UTC instant — provenance of the run, not a floating recording time — so the Clock
Contract's floating rule does not apply here, and this docstring says so as §1 asks.
"""

from __future__ import annotations

import datetime as _dt
import json
import os
import re

import build_id

SCHEMA = "tepna.verdict/1"
STATUSES = ("PASS", "FAIL", "SHORTFALL", "UNDERPOWERED", "NOT_RUN", "NOT_APPLICABLE", "UNKNOWN")
DIRECTIONS = ("lte", "gte", "eq", "within")
SCOPES = ("internal", "publishable")
_SHA = re.compile(r"^[0-9a-f]{7,40}$")
NO_GIT_REASON = "the tree this ran in is not a git checkout (build_id.probe found no sha)"

_commit: str | None | bool = False  # False = not probed yet; None = probed, no .git


def commit_sha(repo_dir: str | None = None) -> str | None:
    """The short sha of the tree this code runs in, probed ONCE (a subprocess); None when git cannot
    say (a tarball deploy) — and the object then carries null, never a placeholder."""
    global _commit
    if _commit is False:
        here = repo_dir or os.path.dirname(os.path.dirname(os.path.abspath(__file__)))
        _commit = build_id.probe(here).get("git")
    return _commit  # type: ignore[return-value]


def make(
    *,
    gate: str,
    status: str,
    population: dict,
    criterion: dict,
    result,
    evidence: list,
    reason: str | None,
    tool: str,
    commit: str | None = None,
    at: str | None = None,
) -> dict:
    """Build a `tepna.verdict/1` object and validate it. Raises ValueError on a malformed one — a gate
    that cannot state its verdict in the shape has not stated it."""
    obj = {
        "schema": SCHEMA,
        "gate": gate,
        "status": status,
        "population": dict(population),
        "criterion": dict(criterion),
        "result": result,
        "evidence": list(evidence),
        "reason": reason,
        "producedBy": _produced_by(tool, commit_sha() if commit is None else commit),
        "at": at or _dt.datetime.now(_dt.timezone.utc).replace(microsecond=0).isoformat().replace("+00:00", "Z"),
        "scope": "internal",
    }
    validate(obj)
    return obj


def _produced_by(tool: str, commit: str | None) -> dict:
    """{tool, commit} — or, when the tree is not a checkout, {tool, commit: null, commitReason}: absence
    says why (§∅), and verdict.js refuses a bare null."""
    if commit is None:
        return {"tool": tool, "commit": None, "commitReason": NO_GIT_REASON}
    return {"tool": tool, "commit": commit}


def unknown(*, gate: str, criterion: dict, evidence: list, tool: str, exc: BaseException) -> dict:
    """A crash is not a verdict: the catch-all shape, naming the exception."""
    return make(
        gate=gate,
        status="UNKNOWN",
        population={"checked": 0, "eligible": 0, "excluded": 0},
        criterion=criterion,
        result=None,
        evidence=evidence,
        reason=f"the gate raised {type(exc).__name__}: {exc}",
        tool=tool,
    )


def validate(obj: dict) -> None:
    """Refuse anything that is not exactly the contract. Mirrors verdict.js."""
    if not isinstance(obj, dict) or obj.get("schema") != SCHEMA:
        raise ValueError(f"schema must be {SCHEMA!r}")
    if not isinstance(obj.get("gate"), str) or not obj["gate"]:
        raise ValueError("gate must be a non-empty string")
    st = obj.get("status")
    if st not in STATUSES:
        raise ValueError(f"status {st!r} is not one of {STATUSES}")
    pop = obj.get("population")
    if not isinstance(pop, dict) or set(pop) != {"checked", "eligible", "excluded"}:
        raise ValueError("population must be exactly {checked, eligible, excluded}")
    for k in ("checked", "eligible", "excluded"):
        if not isinstance(pop[k], int) or isinstance(pop[k], bool) or pop[k] < 0:
            raise ValueError(f"population.{k} must be a non-negative integer")
    if pop["checked"] + pop["excluded"] != pop["eligible"]:
        raise ValueError("population: checked + excluded must equal eligible")
    crit = obj.get("criterion")
    if not isinstance(crit, dict) or set(crit) != {"name", "threshold", "unit", "direction"}:
        raise ValueError("criterion must be exactly {name, threshold, unit, direction}")
    if crit["direction"] not in DIRECTIONS:
        raise ValueError(f"criterion.direction {crit['direction']!r} is not one of {DIRECTIONS}")
    if not isinstance(crit["name"], str) or not crit["name"] or not isinstance(crit["unit"], str):
        raise ValueError("criterion.name must be a non-empty string and criterion.unit a string")
    thr = crit["threshold"]
    if crit["direction"] == "within":
        if not (
            isinstance(thr, list)
            and len(thr) == 2
            and all(isinstance(x, (int, float)) and not isinstance(x, bool) for x in thr)
        ):
            raise ValueError("criterion.threshold for 'within' must be a [lo, hi] pair")
    elif not isinstance(thr, (int, float)) or isinstance(thr, bool):
        raise ValueError("criterion.threshold must be a finite number")
    if obj.get("result") is not None and not isinstance(obj["result"], dict):
        raise ValueError("result must be an object of measured quantities, or null")
    ev = obj.get("evidence")
    if not isinstance(ev, list) or not all(isinstance(e, str) and e for e in ev):
        raise ValueError("evidence must be a list of non-empty strings")
    reason = obj.get("reason")
    if st == "PASS":
        if reason is not None:
            raise ValueError("a PASS carries reason: null")
        if pop["checked"] == 0:
            raise ValueError("a PASS over checked: 0 is the examined-nothing shape")
        if not ev:
            raise ValueError("a PASS needs evidence")
        if obj.get("result") is None:
            raise ValueError("a PASS carries a result")
    elif not isinstance(reason, str) or not reason:
        raise ValueError(f"a {st} requires a reason")
    if st in ("FAIL", "SHORTFALL"):
        if obj.get("result") is None:
            raise ValueError(f"a {st} carries a result — what was measured")
        if not ev:
            raise ValueError(f"a {st} needs evidence")
    if st == "UNDERPOWERED" and not any(ch.isdigit() for ch in str(reason)):
        raise ValueError("an UNDERPOWERED reason names the minimum and the count as numbers")
    if st in ("NOT_RUN", "NOT_APPLICABLE") and obj.get("result") is not None:
        raise ValueError(f"a {st} carries result: null")
    pb = obj.get("producedBy")
    if (
        not isinstance(pb, dict)
        or not {"tool", "commit"} <= set(pb) <= {"tool", "commit", "commitReason"}
        or not isinstance(pb["tool"], str)
        or not pb["tool"]
    ):
        raise ValueError("producedBy must be {tool, commit[, commitReason]}")
    if pb["commit"] is None:
        if not isinstance(pb.get("commitReason"), str) or not pb["commitReason"]:
            raise ValueError("producedBy.commit is null without a commitReason (say why)")
    elif not isinstance(pb["commit"], str) or not _SHA.match(pb["commit"]):
        raise ValueError("producedBy.commit is a 7–40 hex sha or null")
    at = obj.get("at")
    if not isinstance(at, str) or not at.endswith("Z"):
        raise ValueError("at must be a UTC instant ending in Z")
    _dt.datetime.fromisoformat(at.replace("Z", "+00:00"))
    if obj.get("scope") not in SCOPES:
        raise ValueError(f"scope must be one of {SCOPES}")


def write(path: str, obj: dict) -> None:
    """Atomic write beside the report (`<path>.tmp` then rename), after validating."""
    validate(obj)
    tmp = path + ".tmp"
    with open(tmp, "w", encoding="utf-8") as fh:
        json.dump(obj, fh, indent=2)
    os.replace(tmp, path)
